### PR TITLE
cloud: support azure blob storage as external storage backend

### LINKED ISSUE #11731: cloud: support azure blob storage as external storage backend
### Is your feature request related to a problem? Please describe:

The current cloud external storage backend doesn't support azure blob storage. 

### Describe the feature you'd like:

support azure blob storage so that we can backup/restore data to/from it.

### Describe alternatives you've considered:

implement the external storage backend interface for azure blob storage

### PR DESCRIPTION
**Pull request by Leavrth**
### What is changed and how it works?

Close #11731  

What's Changed: implement the external storage backend interface for azure blob storage

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- PR to update `pingcap/tidb-ansible`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- No code

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/cloud/azure/src/azblob.rs [hunk 1]
```diff
@@ -0,0 +1,837 @@
...
+    fn is_retryable(&self) -> bool {
+        matches!(self, Self::TimeOut(_))
+    }
+}
+
+/// Specifies the minimum size to use multi-block upload.
+/// Azure requires each part to be at least 4 MiB compatible with 2016 version.
+/// https://docs.microsoft.com/en-us/rest/api/storageservices/understanding-block-blobs--append-blobs--and-page-blobs
+/// while azure-sdk-for-go use 256 MiB
+const MINIMUM_BLOCK_SIZE: usize = 128 * 1024 * 1024; // 128MB
```
Comment (YuJuncen) [inline components/cloud/azure/src/azblob.rs:260]: When a file reaches 96M, we will split it into multi files. The multi-part upload may never be triggered if we set the block size to 128M.

Comment (Leavrth) [reply components/cloud/azure/src/azblob.rs]: Done. Please take another look.

File: components/cloud/azure/src/azblob.rs [hunk 2]
```diff
@@ -0,0 +1,740 @@
...
+impl ContainerBuilder for TokenCredContainerBuilder {
+    async fn get_client(&self) -> io::Result<Arc<ContainerClient>> {
+        // only the thread get the modify_lock can update the token,
+        // so that this thread can get the token before lock wirte-lock
+        // avoid to block other threads too much time.
+        let mut modify_lock = None;
+        {
+            let token_response = self.token_cache.read().unwrap();
+            if let Some(ref t) = *token_response {
+                let interval = t.0.expires_on - Utc::now();
```
Comment (3pointer) [inline components/cloud/azure/src/azblob.rs:391]: is the expires_on use the utc in any place? what if create azure client in Asia🤔

Comment (Leavrth) [reply components/cloud/azure/src/azblob.rs:391]: Fortunately, the expiration time from the server is the time interval, and the SDK also uses `UTC:: now()`. 
https://github.com/Azure/azure-sdk-for-rust/blob/c5b5e9d86ba8a14586faefaaf8e3cc155c73a24c/sdk/identity/src/token_credentials/client_secret_credentials.rs#L163-L167

File: components/cloud/azure/src/azblob.rs [hunk 3]
```diff
@@ -0,0 +1,740 @@
...
+    tenant_id: String,
+    client_secret: ClientSecret,
+}
+
+impl std::fmt::Debug for CredentialInfo {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("CredentialInfo")
+            .field("client_id", &self.client_id)
+            .field("tenant_id", &self.tenant_id)
+            .field("client_secret", &"?")
```
Comment (kennytm) [inline components/cloud/azure/src/azblob.rs:53]: just use `#[derive(Debug)]`. 

`oauth2::ClientSecret` is a "secret type" and the debug representation is `"ClientSecret([redacted])"` already.

File: components/cloud/azure/src/azblob.rs [hunk 4]
```diff
@@ -0,0 +1,740 @@
...
+            }
+        }
+        None
+    }
+
+    fn load_env_account_name() -> Option<StringNonEmpty> {
+        match env::var(ENV_ACCOUNT_NAME).ok() {
+            Some(name) => StringNonEmpty::opt(name),
+            None => None,
+        }
```
Comment (kennytm) [inline components/cloud/azure/src/azblob.rs:118]:
```suggestion
        env::var(ENV_ACCOUNT_NAME).ok().and_then(StringNonEmpty::opt)
```

File: components/cloud/azure/src/azblob.rs [hunk 5]
```diff
@@ -0,0 +1,740 @@
...
+    }
+
+    fn load_env_shared_key() -> Option<StringNonEmpty> {
+        match env::var(ENV_SHARED_KEY).ok() {
+            Some(key) => StringNonEmpty::opt(key),
+            None => None,
+        }
```
Comment (kennytm) [inline components/cloud/azure/src/azblob.rs:125]: ditto

File: components/cloud/azure/src/azblob.rs [hunk 6]
```diff
@@ -0,0 +1,740 @@
...
+        // only the thread get the modify_lock can update the token,
+        // so that this thread can get the token before lock wirte-lock
+        // avoid to block other threads too much time.
+        let mut modify_lock = None;
+        {
+            let token_response = self.token_cache.read().unwrap();
+            if let Some(ref t) = *token_response {
+                let interval = t.0.expires_on - Utc::now();
+                // keep token updated 5 minutes before it expires
+                if interval > chDuration::minutes(5) {
```
Comment (kennytm) [inline components/cloud/azure/src/azblob.rs:393]: this "5" thing should be a constant

File: components/cloud/azure/src/azblob.rs [hunk 7]
```diff
@@ -0,0 +1,740 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+use async_trait::async_trait;
+use azure_core::auth::{TokenCredential, TokenResponse};
+use azure_core::prelude::*;
+use azure_identity::token_credentials::{ClientSecretCredential, TokenCredentialOptions};
+use azure_storage::blob::prelude::*;
+use azure_storage::core::{prelude::*, ConnectionStringBuilder};
+use chrono::{Duration as chDuration, Utc};
```
Comment (kennytm) [inline components/cloud/azure/src/azblob.rs:8]: The proper casing should be

```suggestion
use chrono::{Duration as ChDuration, Utc};
```

**Review by 3pointer — APPROVED** [review]

**Review by kennytm — APPROVED** [review]

**Pull request merged**